### ISSUE TITLE
Dead containers are not detected in swarm if the host reboots (too fast?)

### ISSUE THREAD

**Issue by SISheogorath** [docker/docker#30327]
**Description**

I had a container in swarm service. And this container was running on *node1*. *node1* is also one of 3 managers. I hard-restarted *node1* which should result in a restart of the container on another host. But instead the host came up again, connected back to swarm and everything looks fine. `docker service ls` shows 1/1 replica and `docker service ps servicename` shows 1 running on *node1* but `docker ps` on *node1* shows up an empty table.

**Steps to reproduce the issue:**
1. create a swarm (I have 3 managers)
2. Start a service without healthcheck with only one instance
3. reboot the host where the instance is running on (in a very fast way?)

**Describe the results you received:**

Everything looks fine...
```
ubuntu@webendcontroller:~$ docker service ls
ID            NAME            MODE        REPLICAS  IMAGE
y1i9kurfwid0  testminecraft4  replicated  1/1       sheogorath/minecraft:latest
```

Even here everything looks nice
```
ID            NAME                  IMAGE                        NODE              DESIRED STATE  CURRENT STATE         ERROR                             PORTS
mmo5do04mcy4  testminecraft4.1      sheogorath/minecraft:latest  serverhoster-2    Running        Running 2 hours ago
```

**BUT**:

```
ubuntu@serverhoster-2:~$ sudo docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES

```

and:

```
ubuntu@serverhoster-2:~$ uptime
 18:42:59 up 5 min,  1 user,  load average: 0.00, 0.01, 0.01
```

**Describe the results you expected:**

I would expect that the containers are rescheduled on restart of a node.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

Manager:
```
ubuntu@webendcontroller:~$ docker version
Client:
 Version:      1.13.0
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Tue Jan 17 09:50:17 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Tue Jan 17 09:50:17 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

The rebooted node:
```
ubuntu@serverhoster-2:~$ sudo docker version
Client:
 Version:      1.13.0
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Tue Jan 17 09:50:17 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   49bf474
 Built:        Tue Jan 17 09:50:17 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

manager:

```ubuntu@webendcontroller:~$ docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 15
Server Version: 1.13.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 39
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: ax68k4qk1bztybpsisoik8c5h
 Is Manager: true
 ClusterID: 8otgo57fymvy1su3lpn7qh70x
 Managers: 3
 Nodes: 4
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.0.4
 Manager Addresses:
  192.168.0.4:2377
  192.168.0.5:2377
  192.168.0.7:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 2f7393a47307a16f8cee44a37b262e8b81021e3e
init version: 949e6fa
Security Options:
 apparmor
Kernel Version: 3.13.0-93-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 1.955 GiB
Name: webendcontroller
ID: 6E6F:KTWN:XDOE:JGNK:LIXX:Y2HR:2VK3:CDAS:SDYN:DIMV:BTFY:7FAR
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

rebooted node
```
ubuntu@serverhoster-2:~$ sudo docker info
Containers: 1
 Running: 0
 Paused: 0
 Stopped: 1
Images: 4
Server Version: 1.13.0
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 18
 Dirperm1 Supported: false
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: active
 NodeID: 1edx41hgs48w3csksrfviz3i5
 Is Manager: true
 ClusterID: 8otgo57fymvy1su3lpn7qh70x
 Managers: 3
 Nodes: 4
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Number of Old Snapshots to Retain: 0
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.0.7
 Manager Addresses:
  192.168.0.4:2377
  192.168.0.5:2377
  192.168.0.7:2377
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 2f7393a47307a16f8cee44a37b262e8b81021e3e
init version: 949e6fa
Security Options:
 apparmor
Kernel Version: 3.13.0-103-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.861 GiB
Name: serverhoster-2
ID: MZYF:NMWZ:XYMN:WS4N:ZRQV:KD56:MROR:YMF4:6HRF:PHE6:QJQD:NEZX
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

I tested it in a OpenStack environment so it was running on VMWare. I can't determine the current version right now but I'll ask the owner about those maybe helpful details.

To restart the machine I used the "Hard Reboot Instance"-Button.

The Image name is `Ubuntu 14.04 - Trusty Tahr - 64-bit - Cloud Based Image - VMDK` Not sure if it's build by our university or an official one. (I guess last but not sure).

By the way I confirmed that it is not important what type of node you reboot. Manager has the problem like workers, too.

My guess it's a time problem to raft can not determine that the node was down.

**Comment by fntlnz** [comment]
@SISheogorath There's only one thing which is not clear, the task is *never* rescheduled on another node or it has no time to do so because the node comes back very fast?

Also `docker service create` does have a few options to play with:

```
docker service create --help | grep restart
      --restart-condition string       Restart when condition is met (none, on-failure, or any)
      --restart-delay value            Delay between restart attempts (default none)
      --restart-max-attempts value     Maximum number of restarts before giving up (default none)
      --restart-window value           Window used to evaluate the restart policy (default none)
```

**Comment by SISheogorath** [comment]
Nope. Never.

```
ubuntu@webendcontroller:~$ docker service ps testminecraft3
ID            NAME                  IMAGE                        NODE            DESIRED STATE  CURRENT STATE          ERROR                             PORTS
lwolv79nuecl  testminecraft3.1      sheogorath/minecraft:latest  serverhoster-2  Running        Running 24 hours ago
```


```
ubuntu@serverhoster-2:~$ sudo docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

**Comment by thaJeztah** [comment]
Could you check if the task is not running on another node? (and it's only a display issue)?
Also interested if the daemon logs (both of the master, and of the node that was rebooted) show anything useful

**Comment by SISheogorath** [comment]
This looks like a problem:
```
time="2017-01-20T18:37:12.940440584Z" level=error msg="failed to start taskManager" error="invalid bind mount source, source path not found: /swarm/volumes/minecraft3" module="node/agent"
```

I guess it's caused by the missing of glusterfs for the mount. Which results in a not starting instance of the container. This would explain why the container is not starting. But it should get redeployed somewhere else.

"Full" log of rebooted node: https://pad.shivering-isles.com/s/Byehq2ZDx

Nothing amazing in the manager logs... :/

"Full" log of the manager node: https://pad.shivering-isles.com/s/HkeDnhbDl

**Comment by thaJeztah** [comment]
ping @cpuguy83 @aaronlehmann any ideas how to resolve this?

**Comment by aaronlehmann** [comment]
ping @stevvooe @LK4D4

It looks like the only thing that happens if `taskManager` fails is that an error is logged: https://github.com/docker/swarmkit/blob/c3d624cfd118584f3d98ba296aa3d1561626247f/agent/worker.go#L376

Does this also need to propagate the error, so that the task failure is reported to the manager?

**Comment by stevvooe** [comment]
@aaronlehmann Starting a `taskManager` must be reliable. The daemon is leaking an error when resolving the task manager, likely at https://github.com/docker/swarmkit/blob/c3d624cfd118584f3d98ba296aa3d1561626247f/agent/exec/controller.go#L103.

Either way, https://github.com/docker/swarmkit/blob/c3d624cfd118584f3d98ba296aa3d1561626247f/agent/worker.go#L402 should propagate the error. I am not seeing why this is getting reported.

It looks like these have something attempting to resolve the mount. I suspect we are just doing too much in the method that resolves the task controller (`executer.Controller`). That should be very lightweight and should only. Looking at https://github.com/docker/docker/blob/master/daemon/cluster/executor/container/container.go#L63, we see that we are validating mounts. https://github.com/docker/docker/blob/d6be0e98027611cfb14a3246ca797bee0936e649/daemon/cluster/executor/container/validate.go#L28 has a `Stat` to the filesystem, which should not exist. Remove that stat and this will work.

**Comment by thaJeztah** [comment]
@stevvooe any suggestion where to move that validation to? (I see i added it in https://github.com/docker/docker/pull/25494)

**Comment by stevvooe** [comment]
@thaJeztah That should be part of starting or stopping the container, at actual mount time.

I reviewed this and missed that detail. No host-specific validation should be done when creating a controller, except for ensuring that the executor can possibly run the task.

**Comment by thaJeztah** [comment]
Hm, that probably requires additional changes, because auto-creating the host directory is the current behavior.

**Comment by stevvooe** [comment]
@thaJeztah One approach would be validate at the start of `Start` or in `Prepare`.

**Comment by stevvooe** [comment]
> Hm, that probably requires additional changes, because auto-creating the host directory is the current behavior.

This is not the behavior for swarm services and we won't be adding because it causes too many issues. One example is when you mean to bind a file that is missing, a directory will be created. This leads to a condition where you need to go delete the directory before adding the file again.

**Comment by thaJeztah** [comment]
You're right about services requiring the host directory to be there; I wasn't sure if the auto-creation is always done for a container (including if that container happens to be started from a "task"), or a separate action on `docker run`

**Comment by stevvooe** [comment]
@thaJeztah Submitted #30732 to move the validation to pre-start.

**Comment by thaJeztah** [comment]
https://github.com/docker/docker/pull/30732 was merged, and should resolve this in 1.13.1, so I'll go ahead and close
